cmd/frontrunning/main.go

```diff
@@ -0,0 +1,198 @@
+package main
+
+import (
+	"bytes"
+	"context"
+	"crypto/ecdsa"
+	"encoding/hex"
+	"encoding/json"
+	"fmt"
+	"log"
+	"math"
+	"math/big"
+	"net/http"
+	"time"
+
+	"github.com/Amenokal-Labs/mev-mantis.git/pkg/utils"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/crypto"
+	"github.com/ethereum/go-ethereum/ethclient"
+	"github.com/ethereum/go-ethereum/ethclient/gethclient"
+	"github.com/ethereum/go-ethereum/rlp"
+	"github.com/ethereum/go-ethereum/rpc"
+)
+
+func main() {
+	ethclient, err := ethclient.Dial("https://goerli.infura.io/v3/" + utils.GetKey("INFURA_KEY"))
+	if err != nil {
+		log.Fatal("[0] ", err)
+
+	}
+
+	rpc, err := rpc.Dial("wss://goerli.infura.io/ws/v3/" + utils.GetKey("INFURA_KEY"))
+	if err != nil {
+		log.Fatal("[1] ", err)
+	}
+	subscriber := gethclient.New(rpc)
+	fmt.Println("client connected..")
+
+	// txs := make(chan *types.Transaction)
+	// _, err = rpcClient.SubscribeFullPendingTransactions(context.Background(), txs)
+	// if err != nil {
+	// 	panic(err)
+	// }
+
+	for {
+		pc, _ := ethclient.PendingTransactionCount(context.Background())
+		fmt.Println("\nPending count:", pc)
+
+		txn, marshalledTxn, from := getTxn(ethclient, subscriber)
+		calldata, contract := getCalldata(txn, marshalledTxn)
+		if contract == "0x" {
+			continue
+		}
+		fmt.Println("    Tx Calldata:", calldata)
+		fmt.Println("     Tx address:", txn.To())
+		fmt.Println("\nContract code:", contract[:26], "...")
+
+		address := from.String()
+		if replaceAddress(calldata, address) != calldata {
+			sendTxn(ethclient, txn)
+		}
+
+		fmt.Println("___________________________")
+		time.Sleep(4 * time.Second)
+	}
+}
+
+func getTxn(_ethclient *ethclient.Client, _subscriber *gethclient.Client) (txn *types.Transaction, marshalledTxn []byte, from common.Address) {
+	hashes := make(chan common.Hash)
+	_, err := _subscriber.SubscribePendingTransactions(context.Background(), hashes)
+	if err != nil {
+		log.Fatal("[2] ", err)
+	}
+	hash := <-hashes
+	fmt.Println("      Tx hash:", hash)
+	txn, _, err = _ethclient.TransactionByHash(context.Background(), hash)
+	if err != nil {
+		log.Fatal("[3] ", err)
+	}
+
+	marshalledTxn, err = txn.MarshalJSON()
+	if err != nil {
+		log.Fatal("[4] ", err)
+	}
+
+	from, err = types.Sender(types.NewLondonSigner(txn.ChainId()), txn)
+	if err != nil {
+		log.Fatal("[5] ", err)
+	}
+
+	return txn, marshalledTxn, from
+}
+
+func getCalldata(_txn *types.Transaction, _marshalledTxn []byte) (calldata string, contract string) {
+	type Contract struct {
+		JsonRpc string `json:"jsonrpc"`
+		Id      int    `json:"id"`
+		Code    string `json:"result"`
+	}
+	to := _txn.To()
```

Comment:
could you add a comment below that explain why you check `to == nil`?

Reply:
Sure
If the tx receiver is nil, it means the tx is a contract deployement (which we're assuming it's a tx with no value to extract, so no point of continuing execution)
![image](https://github.com/Amenokal-Labs/mev-mantis/assets/128217762/38323924-ee63-4a31-b849-51ea793ffb29)
![image](https://github.com/Amenokal-Labs/mev-mantis/assets/128217762/4d1f698a-ceac-48ac-9926-b5837efd16cd)

(screenshots from chapter for of ethereum yellow paper)